src/formats/MMTF.cpp

```diff
@@ -238,172 +248,259 @@ void MMTFFormat::read(Frame& frame) {
             // This is a string in MMTF, differs from the name as then increments linearly
             // For example, the fourth chainid in ( A B A B) would be D, not B (the chainname)
             residue.set("chainid", structure_.chainIdList[chainIndex_]);
-
-            // An integer
             residue.set("chainindex", chainIndex_);
-
-            if (!structure_.chainNameList.empty()) {
+            if (!mmtf::isDefaultValue(structure_.chainNameList)) {
                 residue.set("chainname", structure_.chainNameList[chainIndex_]);
             }
 
             frame.add_residue(std::move(residue));
             groupIndex_++;
         }
-
         chainIndex_++;
     }
-
     modelIndex_++;
+
+    // Add additional global (not by group) bonds
     for (size_t i = 0; i < structure_.bondAtomList.size() / 2; i++) {
         auto atom1 = static_cast<size_t>(structure_.bondAtomList[i * 2]);
         auto atom2 = static_cast<size_t>(structure_.bondAtomList[i * 2 + 1]);
 
         // We are below the atoms we care about
         if (atom1 < atomSkip_ || atom2 < atomSkip_) {
+            if (!(atom1 < atomSkip_ && atom2 < atomSkip_)) {
+                warning("MMTF Reader", "chemfiles can not represent bonds between different models");
+            }
             continue;
         }
 
         // We are above the atoms we care about
         if (atom1 > atomIndex_ || atom2 > atomIndex_) {
+            if (!(atom1 > atomIndex_ && atom2 > atomIndex_)) {
+                warning("MMTF Reader", "chemfiles can not represent bonds between different models");
+            }
             continue;
         }
 
-        size_t atom_idx1 = atom1 - atomSkip_;
-        size_t atom_idx2 = atom2 - atomSkip_;
-
-        frame.add_bond(atom_idx1, atom_idx2);
+        frame.add_bond(atom_id(atom1), atom_id(atom2));
     }
```

Comment:
Note that this is not efficient. I've reworked this code, but never pushed upstream. I'll try to do this shortly.

Reply:
Is the inefficiency due to the fact that all bonds are iterated over, instead of just the ones related to the current model?

Reply:
No, it has to with adding peptide bonds in an odd order. Since the bonds are stored in a `sorted_array`, this causes a lot of moving around memory.

Reply:
Oh, OK. Out of curiosity, what are the gains you have? And if you have an example file where this is an issue, I am also interested. We might be able to change the way we do the sorting in `Connectivity` to help too.